src/services/addressProvider.ts

```diff
@@ -24,10 +24,18 @@ export class AddressProvider<T extends ChainId> extends Service {
     `function addressesMetadataByIdStartsWith(string) public view returns (${AddressMetadataAbi}[] memory)`,
   ];
   private contract: WrappedContract;
+  private cachedAddressesById: Map<
+    ContractAddressId,
+    {
+      address: Address;
+      timestamp: number;
+    }
+  >;
```

Comment:
Each contract address id is mapped to an address and timestamp, so that we can check if the cache is fresh or expired for that particular contract address id.